### DISCUSSION TITLE
How to write tests for a lua filter

### DISCUSSION THREAD

**Discussion by RudolfVonKrugstein** [jgm/pandoc#10323] [Q&A] [answered]
Hi,

I am trying to write tests using `busted` for a lua filter for pandoc, and I am unable to get the tests and using the filter by pandoc to work at the same time.
I think I can break it down to something documented [here](https://pandoc.org/lua-filters.html) not working:

When I do:

```lua
function CodeBlock(el)
  print("code block")
  return el
end
```

It works, by which I mean: I can see the output "code block" when compiling a markdown with a code block using this filter.
But when I do:

```lua
return {
	CodeBlock = function(el)
		print("code block")
		return CodeBlock(el)
	end,
}
```

pandoc does not pick up on `CodeBlock` (I cannot see the output).

But, my problem is that I cannot get the first version to work for testing with busted because I am unable to import the module as a simple `require` does not work.

My pandoc version:
```bash
> pandoc --version
pandoc 3.1.11.1
Features: +server +lua
Scripting engine: Lua 5.4
User data directory: /home/nathan/.local/share/pandoc
Copyright (C) 2006-2023 John MacFarlane. Web: https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.
```

Thanks for any help!

**Comment by jgm** [comment]
Do you want `return pandoc.CodeBlock(el)` ?

**Reply by RudolfVonKrugstein** [reply]
You are correct, my example contained an error. Still, if I write a filter with just this content:

```lua
return {
	CodeBlock = function(el)
		print("code block")
		return pandoc.CodeBlock(el)
	end,
}
```

pandoc does never run the `CodeBlock` function.

**Reply by RudolfVonKrugstein** [reply]
But if I just write

```lua
return {
	{
		CodeBlock = function(el)
			print("code block")
			return el
		end,
	},
}
```

it works! Is the documentation [here](https://pandoc.org/lua-filters.html) incorrect?

**Reply by jgm** [reply]
Check the documentation for `pandoc.CodeBlock`.  The argument is presumably supposed to be some text, not a pandoc element.

**Reply by RudolfVonKrugstein** [reply]
Sorry, I don't seem to be clear.

If I write:

```lua
return {
	CodeBlock = function(el)
		print("code block")
		-- some other code
	end,
}
```

the function never gets called, even so I do have CodeBlocks in my document.

But if I write:

```lua
return {
	{
		CodeBlock = function(el)
			print("code block")
			-- some other code
		end,
	},
}
```

or

```lua
function CodeBlock(el)
  print("code block")
  -- some other code
end
```

The function does get called.
But according to [this](https://pandoc.org/lua-filters.html), all examples (including the first one) should work.

**Reply by jgm** [reply]
The first one will work in pandoc 3.5 but not earlier versions.

**Reply by RudolfVonKrugstein** [reply]
ah ok, thank your for the help!

**Comment by tarleb** [comment, marked as answer]
As an answer to the general question of how to test filters: I asked myself the same thing, and ended up writing [perevir](https://github.com/tarleb/perevir) for that. It's work in progress, but it's already serving me well.

**Reply by RudolfVonKrugstein** [reply]
Uh, thanks!
